File: sdk/python/src/llm_observatory/exporters/otlp.py
"""OTLP HTTP exporter for sending traces to observatory backend."""

from __future__ import annotations

import logging
from typing import Any

import httpx

from llm_observatory.exporters import BaseExporter

logger = logging.getLogger(__name__)


class OTLPExporter(BaseExporter):
    """Exports traces to an OTLP-compatible backend via HTTP."""

    def __init__(
        self,
        endpoint: str = "http://localhost:8000",
        api_key: str | None = None,
        timeout: float = 10.0,
        max_retries: int = 3,
    ) -> None:
        """Initialize the exporter.

        Args:
            endpoint: Backend endpoint URL.
            api_key: Optional API key for authentication.
            timeout: Request timeout in seconds.
            max_retries: Maximum number of retries on failure.
        """
        self.endpoint = endpoint.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout
        self.max_retries = max_retries
        self._client = httpx.Client(timeout=timeout)

    def export(self, spans: list[dict[str, Any]]) -> None:
        """Export spans to the backend.

        Args:
            spans: List of span dictionaries to export.
        """
        headers = {"Content-Type": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"

        for attempt in range(self.max_retries):
            try:
                response = self._client.post(
                    f"{self.endpoint}/v1/traces/batch",
                    json={"spans": spans},
                    headers=headers,
                )
                response.raise_for_status()
                logger.debug(f"Exported {len(spans)} spans successfully")
                return
            except httpx.HTTPStatusError as e:
                logger.warning(
                    f"Export failed (attempt {attempt + 1}/{self.max_retries}): "
                    f"HTTP {e.response.status_code} - {e.response.text}"
                )
                if attempt == self.max_retries - 1:
                    msg = (
                        f"Export failed after {self.max_retries} attempts, "
                        f"dropping {len(spans)} spans"
                    )
                    logger.error(msg)
            except httpx.HTTPError as e:
                logger.warning(
                    f"Export failed (attempt {attempt + 1}/{self.max_retries}): {e}"
                )
                if attempt == self.max_retries - 1:
                    msg = (
                        f"Export failed after {self.max_retries} attempts, "
                        f"dropping {len(spans)} spans"
                    )
                    logger.error(msg)

    def flush(self) -> None:
        """Flush any buffered spans."""
        self._client.close()

    def __del__(self) -> None:
        """Clean up the client."""
        try:
            self._client.close()
        except Exception:
            pass

File: sdk/typescript/src/exporters/otlp.ts
/**
 * OTLP HTTP exporter for sending traces to observatory backend.
 */

import type { Span } from "../tracer";

export interface OTLPExporterOptions {
  endpoint?: string;
  apiKey?: string;
  timeout?: number;
  maxRetries?: number;
}

function spanToDict(span: Span): Record<string, unknown> {
  return {
    id: span.id,
    trace_id: span.traceId,
    parent_id: span.parentId ?? null,
    name: span.name,
    span_type: span.spanType,
    start_time: span.startTime,
    end_time: span.endTime ?? null,
    status: span.status,
    input: span.input ?? null,
    output: span.output ?? null,
    tokens_input: span.tokens?.input ?? null,
    tokens_output: span.tokens?.output ?? null,
    cost_usd: span.costUsd ?? null,
    metadata: span.metadata,
    attributes: span.attributes,
  };
}

export class OTLPExporter {
  private endpoint: string;
  private apiKey?: string;
  private timeout: number;
  private maxRetries: number;
  private buffer: Record<string, unknown>[] = [];

  constructor(options: OTLPExporterOptions = {}) {
    this.endpoint = (options.endpoint || "http://localhost:8000").replace(/\/$/, "");
    this.apiKey = options.apiKey;
    this.timeout = options.timeout ?? 10000;
    this.maxRetries = options.maxRetries ?? 3;
  }

  export(spans: Span[]): void {
    const dicts = spans.map(spanToDict);
    this.buffer.push(...dicts);
    this.flush();
  }

  async flush(): Promise<void> {
    if (this.buffer.length === 0) return;

    const spans = this.buffer.splice(0);
    const headers: Record<string, string> = { "Content-Type": "application/json" };
    if (this.apiKey) {
      headers["Authorization"] = `Bearer ${this.apiKey}`;
    }

    for (let attempt = 0; attempt < this.maxRetries; attempt++) {
      try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeout);

        const response = await fetch(`${this.endpoint}/v1/traces/batch`, {
          method: "POST",
          headers,
          body: JSON.stringify({ spans }),
          signal: controller.signal,
        });

        clearTimeout(timeoutId);

        if (!response.ok) {
          const text = await response.text().catch(() => "");
          console.warn(
            `OTLP export failed (attempt ${attempt + 1}/${this.maxRetries}): ` +
            `HTTP ${response.status} - ${text}`
          );
          if (attempt === this.maxRetries - 1) {
            console.error(`OTLP export failed after ${this.maxRetries} attempts, dropping ${spans.length} spans`);
          }
          continue;
        }

        return;
      } catch (error) {
        console.warn(
          `OTLP export failed (attempt ${attempt + 1}/${this.maxRetries}): ${error}`
        );
        if (attempt === this.maxRetries - 1) {
          console.error(`OTLP export failed after ${this.maxRetries} attempts, dropping ${spans.length} spans`);
        }
      }
    }
  }
}
